include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,307 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht    *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief construct histogram
+ */
+
+#ifndef XTENSOR_HISTOGRAM_HPP
+#define XTENSOR_HISTOGRAM_HPP
+
+#include "xtensor.hpp"
+#include "xsort.hpp"
+
+namespace xt
+{
+
+    /**
+     * @ingroup basic_functions
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param The data.
+     * @param The bin-edges.
+     * @param Weight factors corresponding to each data-point.
+     * @param If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
+    )
+    {
+        // counter type
+        using size_type = detail::common_size_type_t<E1,E2,E3>;
+
+        // basic checks
+        #ifdef XTENSOR_ENABLE_ASSERT
+
+            // - shape and size
+            XTENSOR_ASSERT( data     .dimension() == 1 );
```

Comment:
`XTENSOR_ASSERT` is correctly defined when `XTENSOR_ENABLE_ASSERT` is not defined, so no need to check for this latter.

Reply:
Fair enough, but omitting the `#ifdef` would not allow the preprocessor to get rid of the `for`-loop need to check the `bin_edges` is sorted. Is there an equivalent of `EXPECT_EQ( xt::sort(bin_edges), bin_edges)`?

Reply:
Sorry missed that one. I think you can use `XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend());`